src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php

```diff
@@ -0,0 +1,215 @@
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Config\Builder;
+
+use Symfony\Component\Config\Definition\ArrayNode;
+use Symfony\Component\Config\Definition\BaseNode;
+use Symfony\Component\Config\Definition\BooleanNode;
+use Symfony\Component\Config\Definition\EnumNode;
+use Symfony\Component\Config\Definition\FloatNode;
+use Symfony\Component\Config\Definition\IntegerNode;
+use Symfony\Component\Config\Definition\NodeInterface;
+use Symfony\Component\Config\Definition\NumericNode;
+use Symfony\Component\Config\Definition\PrototypedArrayNode;
+use Symfony\Component\Config\Definition\ScalarNode;
+use Symfony\Component\Config\Definition\StringNode;
+use Symfony\Component\Config\Definition\VariableNode;
+
+/**
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ *
+ * @internal
+ */
+final class ArrayShapeGenerator
+{
+    public const FORMAT_PHPDOC = 'phpdoc';
+    public const FORMAT_JETBRAINS_ATTRIBUTE = 'jetbrains_attribute';
+
+    /**
+     * @param self::FORMAT_* $format
+     */
+    public static function generate(ArrayNode $node, string $format): string
+    {
+        if (self::FORMAT_PHPDOC === $format) {
+            return self::prependPhpDocWithStar(self::doGeneratePhpDoc($node));
+        }
+
+        if (self::FORMAT_JETBRAINS_ATTRIBUTE === $format) {
+            return self::doGenerateJetBrainsArrayShape($node);
+        }
+
+        throw new \LogicException(\sprintf('Unsupported format to generate array shape. Expected one of "%s", got "%s".', implode('", "', [self::FORMAT_PHPDOC, self::FORMAT_JETBRAINS_ATTRIBUTE]), $format));
+    }
+
+    private static function doGeneratePhpDoc(NodeInterface $node, int $nestingLevel = 1): string
+    {
+        if (!$node instanceof ArrayNode) {
+            return $node->getName();
+        }
+
+        if ($node instanceof PrototypedArrayNode) {
+            $prototype = $node->getPrototype();
+            if ($prototype instanceof ArrayNode) {
+                return 'array<'.self::doGeneratePhpDoc($prototype, $nestingLevel).'>';
+            }
+
+            return 'array<'.self::handleNodeType($prototype).'>';
+        }
+
+        if (!($children = $node->getChildren()) && !$node->getParent() instanceof PrototypedArrayNode) {
+            return 'array<array-key, mixed>';
+        }
+
+        $arrayShape = "array{\n";
+
+        $handleNode = function (NodeInterface $node, int $nestingLevel): string {
+            if ($node instanceof ArrayNode) {
+                return self::doGeneratePhpDoc($node, 1 + $nestingLevel);
+            } else {
+                return self::handleNodeType($node);
+            }
+        };
+
+        /** @var NodeInterface $child */
+        foreach ($children as $child) {
+            $arrayShape .= str_repeat(' ', $nestingLevel * 4).self::dumpNodeKey($child).': ';
+
+            if ($child instanceof PrototypedArrayNode) {
```

Comment:
instead of this special case for `PrototypedArrayNode` everywhere, why not making `$handleNode` able to support it (my suggestion of using a private static method will simplify this a lot, as recursive closures are more painful than recursive methods due to needing to capture the closure by reference)